# In racquetball, a player continues to serve as long as she is winning; a point is scored only when a player serves and wins the rally. The first player to 21 points wins the game. Assume that you serve first and have a probability of 0.6 of winning a rally when you serve, and a 0.5 probability of winning a rally when you don't. 

# Estimate, by simulation, the probability that you will win a game

In [1]:
import numpy as np

In [8]:
def rally_you_served((your_score, their_score)):
    rand = np.random.random()
    if rand<0.6:
        your_score+=1
        success = True
    else:
        success = False
    return (your_score, their_score), success

In [9]:
def rally_they_served((your_score, their_score)):
    rand = np.random.random()
    if rand<0.5:
        success = True
    else:
        their_score += 1
        success = False
    return (your_score, their_score), success

In [20]:
def simulate_game():
    (your_score, their_score) = (0,0)
    success = True
    
    while max(your_score, their_score)<=21:
        if max(your_score, their_score)==21:
            break
        
        if success:
            (your_score, their_score), success = rally_you_served((your_score, their_score))
        else:
            (your_score, their_score), success = rally_they_served((your_score, their_score))
            
    return (your_score, their_score)

In [25]:
N_sims = 10000
score = 0

for i in range(N_sims):
    (your_score, their_score) = simulate_game()
    if your_score == 21:
        score += 1

In [26]:
print(float(score)/N_sims)

0.8198


# The probability that you'll win a game is approximately 0.82